server/jdkhttp-server/src/main/scala/sttp/tapir/server/jdkhttp/internal/ParsedMultiPart.scala

```diff
@@ -44,48 +49,125 @@ case class ParsedMultiPart(headers: Map[String, Seq[String]], body: Array[Byte])
-          case AfterHeaderSpace if !line.startsWith(boundary) => state.addBody(line.getBytes())
-          case AfterHeaderSpace                               => state.completePart()
-        }
+      .continually(inputStream.read())
+      .sliding(2)
+      .takeWhile(twoChars => twoChars != Seq('\r', '\n') && twoChars != Seq('\n') && !twoChars.lastOption.contains(-1))
+      .map(_.head.toByte)
+      .toArray
+
+  private def readStringUntilNewline(inputStream: InputStream) = {
+    val bytes = readUntilNewline(inputStream)
+    new String(bytes)
+  }
+
+  def parseMultipartBody(inputStream: InputStream, boundary: String): Seq[ParsedMultiPart] = {
+    val boundaryBytes = boundary.getBytes
+
+    @tailrec
+    def recursivelyParseState(state: ParseData): ParseData = state.parseState match {
+      case Default =>
+        val readString = readStringUntilNewline(inputStream)
+        if (readString == boundary) recursivelyParseState(state.changeState(AfterBoundary))
+        else state.done()
+      case AfterBoundary =>
+        val line = readStringUntilNewline(inputStream)
+        if (line.trim.isBlank) recursivelyParseState(state.changeState(AfterHeaderSpace))
+        else recursivelyParseState(state.addHeader(line))
+      case AfterParsedBody =>
+        val line = readStringUntilNewline(inputStream)
+        if (line.trim.isBlank) state.done()
+        else recursivelyParseState(state.addHeader(line).changeState(AfterBoundary))
+      case AfterHeaderSpace =>
+        val partInputStream = parseMultipartBodyPart(inputStream, boundaryBytes)
+        recursivelyParseState(state.completePart(partInputStream))
+    }
+
+    recursivelyParseState(ParseData(empty, List.empty, Default)).completedParts
+  }
+
+  private val TempFileThreshold = 52_428_800
+  private sealed trait PartStream {
+    val stream: OutputStream
+    def convertToFileIfThresholdMet(numReadBytes: Int): PartStream = this match {
+      case ByteStream(os) if numReadBytes >= TempFileThreshold =>
+        val newFile = createTempFile()
+        val fileOutputStream = new FileOutputStream(newFile)
+        os.writeTo(fileOutputStream)
+        os.close()
+        FileStream(newFile, fileOutputStream)
+      case _ => this
+    }
+    def write(byte: Int): Unit = this.stream.write(byte)
+  }
+  private case class FileStream(tempFile: TapirFile, stream: FileOutputStream) extends PartStream
+  private case class ByteStream(stream: ByteArrayOutputStream = new ByteArrayOutputStream()) extends PartStream
+
+  private def parseMultipartBodyPart(is: InputStream, boundaryBytes: Array[Byte]): InputStream = {
+    @tailrec
+    def recursivelyParseBodyBytes(outputStream: PartStream, lastXBytes: Array[Byte], numReadBytes: Int): InputStream = {
+      val currentByte = is.read()
+      if (currentByte == -1) throw new IllegalArgumentException("Parsing multipart failed, ran out of bytes before finding boundary")
```

Comment:
This isn't needed to pass the tests, but i guess some infinite loop could occur if some invalid part was passed. Is this an okay way to handle that error?

Reply:
Yes, exceptions here are fine. I'm not sure if IAE is the correct one here, though, it's not that an invalid argument was passed, but rather that the body is malformed. Maybe simply a RuntimeException?

Reply:
What would be the resulting error of a RuntimeException?
Wouldn't it make most sense if the resulting error is a 400 or 422 if its because of a malformed body?

Reply:
Hm yeah currently any exception will end up returning a 500. Maybe let's create a separate issue to provide a built-in exception for signalling a specific result, in case it's not possible to communicate this in another way?

Reply:
I'm not sure there is another way? I assume the http4s and others use the failure part of the effect, but the "effect" in `jdkhttp-server` doesn't really offer a way we could handle this as far as i could tell.

Reply:
What we could do is add a `RespondWithException`, which 

(a) would contain the status code and message to return to the user
(b) would be handled by the standard `ExceptionInterceptor`

then we could use it in situations as those